File: pkg/agent/core/ngt/handler/grpc/remove.go
// Copyright (C) 2019-2026 vdaas.org vald team <vald@vdaas.org>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//	https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package grpc

import (
	"context"
	"fmt"

	"github.com/vdaas/vald/apis/grpc/v1/payload"
	"github.com/vdaas/vald/apis/grpc/v1/vald"
	"github.com/vdaas/vald/internal/errors"
	"github.com/vdaas/vald/internal/info"
	"github.com/vdaas/vald/internal/log"
	"github.com/vdaas/vald/internal/net/grpc"
	"github.com/vdaas/vald/internal/net/grpc/codes"
	"github.com/vdaas/vald/internal/net/grpc/errdetails"
	"github.com/vdaas/vald/internal/net/grpc/errhandler"
	"github.com/vdaas/vald/internal/net/grpc/status"
	"github.com/vdaas/vald/internal/observability/attribute"
	"github.com/vdaas/vald/internal/observability/trace"
	"github.com/vdaas/vald/internal/strings"
	"github.com/vdaas/vald/internal/sync"
)

func (s *server) Remove(
	ctx context.Context, req *payload.Remove_Request,
) (res *payload.Object_Location, err error) {
	_, span := trace.StartSpan(ctx, apiName+"/"+vald.RemoveRPCName)
	defer trace.End(span)
	id := req.GetId()
	uuid := id.GetId()
	if err = s.validateUUID(span, vald.RemoveRPCName, ngtResourceType+"/ngt.Remove",
		uuid, req); err != nil {
		return nil, err
	}
	err = s.ngt.DeleteWithTime(uuid, req.GetConfig().GetTimestamp())
	if err != nil {
		var attrs []attribute.KeyValue
		if errors.Is(err, errors.ErrFlushingIsInProgress) {
			err = status.WrapWithAborted("Remove API aborted to process remove request due to flushing indices is in progress", err,
				&errdetails.RequestInfo{
					RequestId:   uuid,
					ServingData: errdetails.Serialize(req),
				},
				s.resourceInfo(ngtResourceType+"/ngt.Remove"))
			log.Warn(err)
			attrs = trace.StatusCodeAborted(err.Error())
		} else if errors.Is(err, errors.ErrObjectIDNotFound(uuid)) {
			err = status.WrapWithNotFound(fmt.Sprintf("Remove API uuid %s not found", uuid), err,
				&errdetails.RequestInfo{
					RequestId:   uuid,
					ServingData: errdetails.Serialize(req),
				},
				s.resourceInfo(ngtResourceType+"/ngt.Remove"))
			log.Warn(err)
			attrs = trace.StatusCodeNotFound(err.Error())
		} else if errors.Is(err, errors.ErrUUIDNotFound(0)) {
			err = status.WrapWithInvalidArgument(fmt.Sprintf("Remove API invalid argument for uuid \"%s\" detected", uuid), err,
				&errdetails.RequestInfo{
					RequestId:   uuid,
					ServingData: errdetails.Serialize(req),
				},
				&errdetails.BadRequest{
					FieldViolations: []*errdetails.BadRequestFieldViolation{
						{
							Field:       "uuid",
							Description: err.Error(),
						},
					},
				},
				s.resourceInfo(ngtResourceType+"/ngt.Remove"))
			log.Warn(err)
			attrs = trace.StatusCodeInvalidArgument(err.Error())
		} else {
			err = status.WrapWithInternal("Remove API failed", err,
				&errdetails.RequestInfo{
					RequestId:   uuid,
					ServingData: errdetails.Serialize(req),
				},
				s.resourceInfo(ngtResourceType+"/ngt.Remove"), info.Get())
			log.Error(err)
			attrs = trace.StatusCodeInternal(err.Error())
		}
		errhandler.RecordSpanAttrs(span, attrs, err)
		return nil, err
	}
	return s.newLocation(uuid), nil
}

func (s *server) StreamRemove(stream vald.Remove_StreamRemoveServer) (err error) {
	ctx, span := trace.StartSpan(stream.Context(), apiName+"/"+vald.StreamRemoveRPCName)
	defer trace.End(span)
	err = grpc.BidirectionalStream(ctx, stream, s.streamConcurrency,
		func(ctx context.Context, req *payload.Remove_Request) (*payload.Object_StreamLocation, error) {
			ctx, sspan := trace.StartSpan(ctx, apiName+"/"+vald.StreamRemoveRPCName+"/id-"+req.GetId().GetId())
			defer trace.End(sspan)
			res, err := s.Remove(ctx, req)
			if err != nil {
				st, _ := status.FromError(err)
				errhandler.RecordSpanStatus(sspan, st, err)
				return &payload.Object_StreamLocation{
					Payload: &payload.Object_StreamLocation_Status{
						Status: st.Proto(),
					},
				}, err
			}
			return &payload.Object_StreamLocation{
				Payload: &payload.Object_StreamLocation_Location{
					Location: res,
				},
			}, nil
		})
	if err != nil {
		st, _ := status.FromError(err)
		errhandler.RecordSpanStatus(span, st, err)
		return err
	}
	return nil
}

func (s *server) MultiRemove(
	ctx context.Context, reqs *payload.Remove_MultiRequest,
) (res *payload.Object_Locations, err error) {
	_, span := trace.StartSpan(ctx, apiName+"/"+vald.MultiRemoveRPCName)
	defer trace.End(span)
	uuids := make([]string, 0, len(reqs.GetRequests()))
	for _, req := range reqs.GetRequests() {
		uuids = append(uuids, req.GetId().GetId())
	}
	err = s.ngt.DeleteMultiple(uuids...)
	if err != nil {
		var attrs []attribute.KeyValue
		if errors.Is(err, errors.ErrFlushingIsInProgress) {
			err = status.WrapWithAborted("MultiRemove API aborted to process remove request due to flushing indices is in progress", err,
				&errdetails.RequestInfo{
					RequestId:   strings.Join(uuids, ", "),
					ServingData: errdetails.Serialize(reqs),
				},
				s.resourceInfo(ngtResourceType+"/ngt.MultiRemove"))
			log.Warn(err)
			attrs = trace.StatusCodeAborted(err.Error())
		} else if notFoundIDs := func() []string {
			aids := make([]string, 0, len(uuids))
			for _, id := range uuids {
				if errors.Is(err, errors.ErrObjectIDNotFound(id)) {
					aids = append(aids, id)
				}
			}
			return aids
		}(); len(notFoundIDs) != 0 {
			err = status.WrapWithNotFound(fmt.Sprintf("MultiRemove API uuids %v not found", notFoundIDs), err,
				&errdetails.RequestInfo{
					RequestId:   strings.Join(uuids, ", "),
					ServingData: errdetails.Serialize(reqs),
				},
				s.resourceInfo(ngtResourceType+"/ngt.MultiRemove"))
			log.Warn(err)
			attrs = trace.StatusCodeNotFound(err.Error())
		} else if errors.Is(err, errors.ErrUUIDNotFound(0)) {
			err = status.WrapWithInvalidArgument(fmt.Sprintf("MultiRemove API invalid argument for uuids \"%v\" detected", uuids), err,
				&errdetails.RequestInfo{
					RequestId:   strings.Join(uuids, ", "),
					ServingData: errdetails.Serialize(reqs),
				},
				&errdetails.BadRequest{
					FieldViolations: []*errdetails.BadRequestFieldViolation{
						{
							Field:       "uuid",
							Description: err.Error(),
						},
					},
				},
				s.resourceInfo(ngtResourceType+"/ngt.MultiRemove"))
			log.Warn(err)
			attrs = trace.StatusCodeInvalidArgument(err.Error())
		} else {
			err = status.WrapWithInternal("MultiRemove API failed", err,
				&errdetails.RequestInfo{
					RequestId:   strings.Join(uuids, ", "),
					ServingData: errdetails.Serialize(reqs),
				},
				s.resourceInfo(ngtResourceType+"/ngt.MultiRemove"), info.Get())
			log.Error(err)
			attrs = trace.StatusCodeInternal(err.Error())
		}
		errhandler.RecordSpanAttrs(span, attrs, err)
		return nil, err
	}
	return s.newLocations(uuids...), nil
}

func (s *server) RemoveByTimestamp(
	ctx context.Context, req *payload.Remove_TimestampRequest,
) (locs *payload.Object_Locations, errs error) {
	ctx, span := trace.StartSpan(ctx, apiName+"/"+vald.RemoveByTimestampRPCName)
	defer trace.End(span)

	var mu, emu sync.Mutex
	locs = new(payload.Object_Locations)

	timestampOpFn := timestampOpsFunc(req.GetTimestamps())
	s.ngt.ListObjectFunc(ctx, func(uuid string, oid uint32, timestamp int64) bool {
		if !timestampOpFn(timestamp) {
			return true
		}
		res, err := s.Remove(ctx, &payload.Remove_Request{
			Id: &payload.Object_ID{
				Id: uuid,
			},
		})
		if err != nil {
			emu.Lock()
			errs = errors.Join(errs, err)
			emu.Unlock()
		}
		if res != nil {
			mu.Lock()
			locs.Locations = append(locs.Locations, res)
			mu.Unlock()
		}
		return true
	})
	if errs != nil {
		st, _ := status.FromError(errs)
		log.Error(errs)
		errhandler.RecordSpanStatus(span, st, errs)
		return nil, errs
	}
	if locs == nil || len(locs.GetLocations()) == 0 {
		err := status.WrapWithNotFound(
			vald.RemoveByTimestampRPCName+" API remove target not found", errors.ErrIndexNotFound,
			&errdetails.RequestInfo{
				ServingData: errdetails.Serialize(req),
			},
			s.resourceInfo(ngtResourceType+"/ngt.Remove"),
		)
		log.Error(err)
		return errhandler.HandleError[payload.Object_Locations](span, codes.NotFound, err)
	}
	return locs, nil
}

func timestampOpsFunc(ts []*payload.Remove_Timestamp) func(int64) bool {
	fns := make([]func(int64) bool, 0, len(ts))
	for _, t := range ts {
		fns = append(fns, timestampOpFunc(t))
	}
	return func(t int64) bool {
		for _, fn := range fns {
			if !fn(t) {
				return false
			}
		}
		return true
	}
}

func timestampOpFunc(ts *payload.Remove_Timestamp) func(int64) bool {
	switch ts.GetOperator() {
	case payload.Remove_Timestamp_EQ:
		return func(t int64) bool {
			return ts.GetTimestamp() == t
		}
	case payload.Remove_Timestamp_NE:
		return func(t int64) bool {
			return ts.GetTimestamp() != t
		}
	case payload.Remove_Timestamp_GE:
		return func(t int64) bool {
			return ts.GetTimestamp() <= t
		}
	case payload.Remove_Timestamp_GT:
		return func(t int64) bool {
			return ts.GetTimestamp() < t
		}
	case payload.Remove_Timestamp_LE:
		return func(t int64) bool {
			return ts.GetTimestamp() >= t
		}
	case payload.Remove_Timestamp_LT:
		return func(t int64) bool {
			return ts.GetTimestamp() > t
		}
	default:
		return func(timestamp int64) bool {
			return false
		}
	}
}

File: rust/bin/agent/src/handler/remove.rs
//
// Copyright (C) 2019-2026 vdaas.org vald team <vald@vdaas.org>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
use algorithm::{Error, MultiError};
use log::{error, info, warn};
use prost::Message;
use proto::{
    payload::v1::{object, remove},
    vald::v1::remove_server,
};
use std::sync::Arc;
use tokio::sync::RwLock;
use tonic::{Code, Status};
use tonic_types::StatusExt;

use super::common::{bidirectional_stream, build_error_details};

async fn remove<S: algorithm::ANN>(
    s: Arc<RwLock<S>>,
    resource_type: &str,
    api_name: &str,
    name: &str,
    ip: &str,
    request: &remove::Request,
) -> Result<object::Location, Status> {
    let _config = match request.config.clone() {
        Some(cfg) => cfg,
        None => return Err(Status::invalid_argument("Missing configuration in request")),
    };
    let id = match request.id.clone() {
        Some(id) => id,
        None => return Err(Status::invalid_argument("Missing ID in request")),
    };
    let uuid = id.id;
    {
        let mut s = s.write().await;
        if uuid.len() == 0 {
            let err = Error::InvalidUUID { uuid: uuid.clone() };
            let resource_type = format!("{}/qbg.Remove", resource_type);
            let resource_name = format!("{}: {}({})", api_name, name, ip);
            let err_details = build_error_details(
                err,
                &uuid,
                request.encode_to_vec(),
                &resource_type,
                &resource_name,
                Some("uuid"),
            );
            let status = Status::with_error_details(
                Code::InvalidArgument,
                format!("Remove API invalid argument for uuid \"{}\" detected", uuid),
                err_details,
            );
            warn!("{:?}", status);
            return Err(status);
        }
        let result = s.remove(uuid.clone()).await;
        match result {
            Err(err) => {
                let resource_type = format!("{}/qbg.Remove", resource_type);
                let resource_name = format!("{}: {}({})", api_name, name, ip);
                let err_msg = err.to_string();
                let mut err_details = build_error_details(
                    err_msg.clone(),
                    &uuid,
                    request.encode_to_vec(),
                    &resource_type,
                    &resource_name,
                    None,
                );
                let status = match err {
                    Error::FlushingIsInProgress {} => {
                        let err_details = build_error_details(
                            err,
                            &uuid,
                            request_bytes,
                            &resource_type,
                            &resource_name,
                            None,
                        );
                        let status = Status::with_error_details(
                            Code::Aborted,
                            "Remove API aborted to process remove request due to flushing indices is in progress",
                            err_details,
                        );
                        warn!("{:?}", status);
                        status
                    }
                    Error::ObjectIDNotFound { uuid: _ } => {
                        let status = Status::with_error_details(
                            Code::NotFound,
                            format!("Remove API uuid {} not found", uuid),
                            err_details,
                        );
                        warn!("{:?}", status);
                        status
                    }
                    Error::UUIDNotFound { uuid: _ } => {
                        err_details
                            .set_bad_request(vec![tonic_types::FieldViolation::new("id", err_msg)]);
                        let status = Status::with_error_details(
                            Code::InvalidArgument,
                            format!("Remove API invalid argument for uuid \"{}\" detected", uuid),
                            err_details,
                        );
                        warn!("{:?}", status);
                        status
                    }
                    _ => {
                        let status = Status::with_error_details(
                            Code::Internal,
                            "Remove API failed",
                            err_details,
                        );
                        error!("{:?}", status);
                        status
                    }
                };
                Err(status)
            }
            Ok(()) => Ok(object::Location {
                name: name.to_owned(),
                uuid: uuid,
                ips: vec![ip.to_owned()],
            }),
        }
    }
}

#[tonic::async_trait]
impl<S: algorithm::ANN + 'static> remove_server::Remove for super::Agent<S> {
    async fn remove(
        &self,
        request: tonic::Request<remove::Request>,
    ) -> std::result::Result<tonic::Response<object::Location>, tonic::Status> {
        info!("Recieved a request from {:?}", request.remote_addr());
        let request = request.get_ref();
        let s = self.s.clone();
        let resource_type = self.resource_type.clone();
        let name = self.name.clone();
        let ip = self.ip.clone();
        let api_name = self.api_name.clone();
        match remove(s, &resource_type, &api_name, &name, &ip, request).await {
            Ok(location) => Ok(tonic::Response::new(location)),
            Err(e) => Err(e),
        }
    }

    #[doc = " A method to remove an indexed vector based on timestamp.\n"]
    async fn remove_by_timestamp(
        &self,
        request: tonic::Request<remove::TimestampRequest>,
    ) -> std::result::Result<tonic::Response<object::Locations>, tonic::Status> {
        info!("Recieved a request from {:?}", request.remote_addr());
        let req = request.get_ref();
        let timestamps = &req.timestamps;

        let mut locations: Vec<object::Location> = Vec::new();
        let mut errors: Vec<Status> = Vec::new();

        // Build timestamp filter function
        let timestamp_filter = |obj_ts: i64| -> bool {
            for ts in timestamps {
                let op = remove::timestamp::Operator::try_from(ts.operator)
                    .unwrap_or(remove::timestamp::Operator::Eq);
                let matches = match op {
                    remove::timestamp::Operator::Eq => obj_ts == ts.timestamp,
                    remove::timestamp::Operator::Ne => obj_ts != ts.timestamp,
                    remove::timestamp::Operator::Ge => obj_ts >= ts.timestamp,
                    remove::timestamp::Operator::Gt => obj_ts > ts.timestamp,
                    remove::timestamp::Operator::Le => obj_ts <= ts.timestamp,
                    remove::timestamp::Operator::Lt => obj_ts < ts.timestamp,
                };
                if !matches {
                    return false;
                }
            }
            true
        };

        // Collect UUIDs to remove based on timestamp filter
        let uuids_to_remove: Vec<String>;
        {
            let s = self.s.read().await;
            let mut matching_uuids = Vec::new();
            s.list_object_func(|uuid, _vec, ts| {
                if timestamp_filter(ts) {
                    matching_uuids.push(uuid);
                }
                true
            })
            .await;
            uuids_to_remove = matching_uuids;
        }

        // Remove each matching object
        for uuid in uuids_to_remove {
            let remove_req = remove::Request {
                id: Some(object::Id { id: uuid.clone() }),
                config: None,
            };
            match remove(
                self.s.clone(),
                &self.resource_type,
                &self.api_name,
                &self.name,
                &self.ip,
                &remove_req,
            )
            .await
            {
                Ok(loc) => locations.push(loc),
                Err(e) => errors.push(e),
            }
        }

        if !errors.is_empty() && locations.is_empty() {
            // All removals failed
            return Err(errors.into_iter().next().unwrap());
        }

        if locations.is_empty() {
            let resource_type = format!("{}/qbg.RemoveByTimestamp", self.resource_type);
            let resource_name = format!("{}: {}({})", self.api_name, self.name, self.ip);
            let err = Error::IndexNotFound {};
            let err_details = build_error_details(
                err,
                "",
                req.encode_to_vec(),
                &resource_type,
                &resource_name,
                None,
            );
            let status = Status::with_error_details(
                Code::NotFound,
                "RemoveByTimestamp API remove target not found",
                err_details,
            );
            error!("{:?}", status);
            return Err(status);
        }

        Ok(tonic::Response::new(object::Locations { locations }))
    }

    #[doc = " Server streaming response type for the StreamRemove method."]
    type StreamRemoveStream = crate::stream_type!(object::StreamLocation);

    #[doc = " A method to remove multiple indexed vectors by bidirectional streaming.\n"]
    async fn stream_remove(
        &self,
        request: tonic::Request<tonic::Streaming<remove::Request>>,
    ) -> std::result::Result<tonic::Response<Self::StreamRemoveStream>, tonic::Status> {
        info!(
            "Received stream remove request from {:?}",
            request.remote_addr()
        );

        let s = self.s.clone();
        let resource_type = self.resource_type.clone() + "/qbg.StreamRemove";
        let name = self.name.clone();
        let ip = self.ip.clone();
        let api_name = self.api_name.clone();

        let process_fn = move |req: remove::Request| {
            let s = s.clone();
            let resource_type = resource_type.clone();
            let name = name.clone();
            let ip = ip.clone();
            let api_name = api_name.clone();
            async move {
                match remove(s, &resource_type, &api_name, &name, &ip, &req).await {
                    Ok(location) => Ok(object::StreamLocation {
                        payload: Some(object::stream_location::Payload::Location(location)),
                    }),
                    Err(status) => Err(status),
                }
            }
        };

        bidirectional_stream(request, self.stream_concurrency, process_fn).await
    }

    #[doc = " A method to remove multiple indexed vectors in a single request.\n"]
    async fn multi_remove(
        &self,
        request: tonic::Request<remove::MultiRequest>,
    ) -> std::result::Result<tonic::Response<object::Locations>, tonic::Status> {
        info!("Recieved a request from {:?}", request.remote_addr());
        let mreq = request.get_ref();
        let uuids: Vec<String> = mreq
            .requests
            .clone()
            .into_iter()
            .filter_map(|x| match x.id {
                Some(id) => Some(id.id),
                None => None,
            })
            .collect();
        {
            let mut s = self.s.write().await;
            let result = s.remove_multiple(uuids.clone()).await;
            match result {
                Err(err) => {
                    let resource_type = self.resource_type.clone() + "/qbg.MultiRemove";
                    let resource_name = format!("{}: {}({})", self.api_name, self.name, self.ip);
                    let request_bytes = mreq.encode_to_vec();
                    let status = match err {
                        Error::FlushingIsInProgress {} => {
                            let err_details = build_error_details(
                                err,
                                &uuids.join(","),
                                request_bytes,
                                &resource_type,
                                &resource_name,
                                None,
                            );
                            let status = Status::with_error_details(
                                Code::Aborted,
                                "MultiRemove API aborted to process remove request due to flushing indices is in progress",
                                err_details,
                            );
                            warn!("{:?}", status);
                            status
                        }
                        Error::ObjectIDNotFound { ref uuid } => {
                            let err_details = build_error_details(
                                &err,
                                uuid,
                                request_bytes,
                                &resource_type,
                                &resource_name,
                                None,
                            );
                            let uuids = Error::split_uuids(uuid.to_string());
                            let status = Status::with_error_details(
                                Code::NotFound,
                                format!("MultiRemove API uuids {:?} not found", uuids),
                                err_details,
                            );
                            warn!("{:?}", status);
                            status
                        }
                        Error::UUIDNotFound { uuid: _ } => {
                            let err_details = build_error_details(
                                err,
                                &uuids.join(","),
                                request_bytes,
                                &resource_type,
                                &resource_name,
                                Some("uuid"),
                            );
                            let status = Status::with_error_details(
                                Code::InvalidArgument,
                                format!(
                                    "MultiRemove API invalid argument for uuids \"{:?}\" detected",
                                    uuids
                                ),
                                err_details,
                            );
                            warn!("{:?}", status);
                            status
                        }
                        _ => {
                            let err_details = build_error_details(
                                err,
                                &uuids.join(","),
                                request_bytes,
                                &resource_type,
                                &resource_name,
                                None,
                            );
                            let status = Status::with_error_details(
                                Code::Internal,
                                "MultiRemove API failed",
                                err_details,
                            );
                            error!("{:?}", status);
                            status
                        }
                    };
                    Err(status)
                }
                Ok(()) => Ok(tonic::Response::new(object::Locations {
                    locations: uuids
                        .iter()
                        .map(|x| object::Location {
                            name: self.name.clone(),
                            uuid: x.to_string(),
                            ips: vec![self.ip.clone()],
                        })
                        .collect(),
                })),
            }
        }
    }
}
